### PR TITLE
`--filter` argument for `split`

### PR DESCRIPTION
**Pull request by FelipeLema**
Be able to pass the bytes and filename to a child shell process.
- start a shell process with the provided command-as-a-string
- forward filename (xaa, xab, xac... ) as $FILE env var
- send bytes to each shell process

### CODE REVIEW

File: src/uu/split/src/split.rs [hunk 1]
```diff
@@ -335,18 +421,27 @@ fn split(settings: &Settings) -> i32 {
                 .as_ref(),
             );
             filename.push_str(settings.additional_suffix.as_ref());
+            // aquí "apunto" $FILE
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:424]: Please remove this comment :)

File: src/uu/split/src/split.rs [hunk 2]
```diff
@@ -335,18 +421,27 @@ fn split(settings: &Settings) -> i32 {
...
             }
             fileno += 1;
-            writer = BufWriter::new(Box::new(
-                OpenOptions::new()
-                    .write(true)
-                    .create(true)
-                    .open(Path::new(&filename))
-                    .unwrap(),
-            ) as Box<dyn Write>);
+            // aquí if … escritor_a_stdin de un proceso ql … else … esta weá
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:430]: please translate this comment!

File: tests/by-util/test_split.rs [hunk 3]
```diff
@@ -156,3 +161,21 @@ fn test_split_additional_suffix() {
...
+    let name = "filtered";
+    let glob = Glob::new(&at, ".", r"x[[:alpha:]][[:alpha:]]$");
+    let n_lines = 3;
+    RandomFile::new(&at, name).add_lines(n_lines);
+
+    // change all characters to 'i'
+    ucmd.args(&["--filter=sed s/./i/g > $FILE", name])
+        .succeeds();
+    // assert all characters are 'i' (command succeded)
+    assert!(glob.collate().iter().find(|&&c| c == ('i' as u8)));
```
Comment (sylvestre) [inline tests/by-util/test_split.rs:179]: please fix this error !

File: src/uu/split/src/split.rs [hunk 4]
```diff
@@ -138,6 +147,12 @@ size is 1000, and default PREFIX is 'x'. With no INPUT, or when INPUT is
     settings.input = input;
     settings.prefix = prefix;
 
+    settings.filter = if let Some(filter) = matches.opt_str("filter") {
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:150]: are you sure there isn't a better way to do that?
like with unwrap_or_else ?

File: src/uu/split/src/split.rs [hunk 5]
```diff
@@ -47,6 +49,12 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
         "additional suffix to append to output file names",
         "SUFFIX",
     );
+    opts.optopt(
+        "",
+        "filter",
+        "write to shell COMMANDl file name is $FILE",
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:55]:
```suggestion
        "write to shell COMMAND file name is $FILE",
```

**Review by sylvestre — CHANGES_REQUESTED** [review]

File: src/uu/split/src/split.rs (hunk 1 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:424]: oh, crap

File: src/uu/split/src/split.rs (hunk 2 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:430]: sorry, it's a mid-dev marker

File: tests/by-util/test_split.rs (hunk 3 above)
Comment (FelipeLema) [reply tests/by-util/test_split.rs]: fixed, although I feel it's not idiomatic the way it turned out

File: src/uu/split/src/split.rs (hunk 1 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:424]: sorry, it's a mid-dev marker

File: src/uu/split/src/split.rs [hunk 6]
```diff
@@ -288,6 +301,76 @@ fn num_prefix(i: usize, width: usize) -> String {
...
+impl Write for FilterWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        self.shell_process
+            .stdin
+            .as_mut()
+            .expect("failed to get shell stdin")
+            .write(buf)
+    }
+    fn flush(&mut self) -> Result<()> {
+        self.shell_process
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:319]: codecov complains here, I guess it is too hard to test?

File: src/uu/split/src/split.rs [hunk 7]
```diff
@@ -288,6 +301,76 @@ fn num_prefix(i: usize, width: usize) -> String {
...
+        let shell_process = Command::new(shell_command)
+            .arg("-c")
+            .arg(command)
+            .stdin(Stdio::piped())
+            .spawn()
+            .expect("Couldn't spawn filter command");
+
+        // restore previous $FILE
+        if let Ok(prev_file) = previous_file_env {
+            env::set_var("FILE", prev_file)
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:352]: is it possible to test this case?

File: src/uu/split/src/split.rs (hunk 4 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs]: replaced

File: src/uu/split/src/split.rs (hunk 7 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:352]: will do

File: src/uu/split/src/split.rs (hunk 6 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:319]: I can `flush()` in `FilterWriter.drop()`

**Comment by FelipeLema** [comment]
still missing windows tests and a test with a failed command

**Comment by sylvestre** [comment]
I guess you noticed that Windows is failing

**Comment by FelipeLema** [comment]
well, right now I'm confused how to write `sh -c 'cat > $FILE'` in windows-esque

Is it `cmd /c print /v "" > "%FILE%"` ? (see discussion [here](https://superuser.com/questions/853580/real-windows-equivalent-to-cat-stdin))

**Comment by FelipeLema** [comment]
OH MY DAYS 

Handling `cmd` arguments has been a problem [for a while](https://github.com/rust-lang/rust/issues/29494#issuecomment-439726937)

I'll change this to a "non-windows"-only feature, then.

File: src/uu/split/src/split.rs [hunk 8]
```diff
@@ -47,6 +52,12 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
         "additional suffix to append to output file names",
         "SUFFIX",
     );
+    opts.optopt(
+        "",
+        "filter",
+        "write to shell COMMAND file name is $FILE",
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:58]:
```suggestion
        "write to shell COMMAND file name is $FILE (Currently not implemented for Windows)",
```

File: src/uu/split/src/split.rs [hunk 9]
```diff
@@ -138,6 +150,13 @@ size is 1000, and default PREFIX is 'x'. With no INPUT, or when INPUT is
     settings.input = input;
     settings.prefix = prefix;
 
+    settings.filter = matches.opt_str("filter");
+
+    if settings.filter.is_some() && cfg!(windows) {
+        // see https://github.com/rust-lang/rust/issues/29494
+        crash!(1, "--filter is not supported in this platform")
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:157]: Would it be possible to use show_error and return instead of crash?

File: src/uu/split/src/split.rs [hunk 10]
```diff
@@ -288,6 +309,134 @@ fn num_prefix(i: usize, width: usize) -> String {
...
+/// A writer that writes to a shell_process' stdin
+struct FilterWriter {
+    /// Running shell process
+    shell_process: Child,
+}
+
+#[cfg(unix)]
+impl Write for FilterWriter {
+    fn write(&mut self, buf: &[u8]) -> Result<usize> {
+        self.shell_process
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:322]: could you add a comment why we are using a shell here?

File: src/uu/split/src/split.rs [hunk 11]
```diff
@@ -288,6 +309,134 @@ fn num_prefix(i: usize, width: usize) -> String {
...
+
+/// Have an environment variable set at a value during this lifetime
+struct WithEnvVarSet {
+    /// Env var key
+    _previous_var_key: String,
+    /// Previous value set to this key
+    _previous_var_value: std::result::Result<String, env::VarError>,
+}
+impl WithEnvVarSet {
+    fn new(key: &str, value: &str) -> WithEnvVarSet {
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:345]: please fix this warning

File: tests/by-util/test_split.rs [hunk 12]
```diff
@@ -4,6 +4,7 @@ extern crate regex;
 use self::rand::{thread_rng, Rng};
 use self::regex::Regex;
 use crate::common::util::*;
+use std::env;
```
Comment (sylvestre) [inline tests/by-util/test_split.rs:7]: please fix this warning too

File: tests/by-util/test_split.rs [hunk 13]
```diff
@@ -156,3 +164,63 @@ fn test_split_additional_suffix() {
...
+#[test]
+#[cfg(unix)]
+fn test_filter_command_fails() {
+    let (at, mut ucmd) = at_and_ucmd!();
+    let name = "filter-will-fail";
+    RandomFile::new(&at, name).add_lines(4);
+
+    assert!(
+        !ucmd
+            .args(&["--filter=/a/path/that/totally/does/not/exist", name])
```
Comment (sylvestre) [inline tests/by-util/test_split.rs:222]: nice :)

File: tests/by-util/test_split.rs [hunk 14]
```diff
@@ -156,3 +164,63 @@ fn test_split_additional_suffix() {
...
+            .fails()
```
Comment (sylvestre) [inline tests/by-util/test_split.rs:223]: fails should be enough ?

fails.().success is funny :)

File: tests/common/util.rs [hunk 15]
```diff
@@ -72,8 +72,12 @@ pub fn repeat_str(s: &str, n: u32) -> String {
 pub struct CmdResult {
     //tmpd is used for convenience functions for asserts against fixtures
     tmpd: Option<Rc<TempDir>>,
+    /// zero-exit from running the Command?
```
Comment (sylvestre) [inline tests/common/util.rs:75]: thanks for the comments!

File: src/uu/split/src/split.rs (hunk 9 above)
Comment (sylvestre) [inline src/uu/split/src/split.rs:157]:
```suggestion
        crash!(1, "--filter is currently not supported in this platform")
```

File: src/uu/split/src/split.rs [hunk 16]
```diff
@@ -47,6 +52,12 @@ pub fn uumain(args: impl uucore::Args) -> i32 {
         "additional suffix to append to output file names",
         "SUFFIX",
     );
+    opts.optopt(
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:55]: if you are motivated, don't hesitate to port this to clap :) (in a different PR)

**Review by sylvestre — COMMENTED** [review]
Looks good. Thanks!
Just some minor comments

File: src/uu/split/src/split.rs (hunk 9 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:157]: will do

Comment (sylvestre) [reply src/uu/split/src/split.rs:157]: And don't hesitate to open an issue on this project as good first bug for potential contributor to implement this on Windows too :)

File: src/uu/split/src/split.rs (hunk 10 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:322]: will add it in `struct FilterWriter` 's docstring

File: src/uu/split/src/split.rs (hunk 9 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs]: manually updated this on

File: src/uu/split/src/split.rs (hunk 16 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:56]: in next PR

File: tests/by-util/test_split.rs (hunk 14 above)
Comment (FelipeLema) [reply tests/by-util/test_split.rs:223]: then I don't `assert!` ? (there's an assertion in `fails()` anyway)

File: src/uu/split/src/split.rs [hunk 17]
```diff
@@ -288,6 +311,142 @@ fn num_prefix(i: usize, width: usize) -> String {
...
+    fn new(key: &str, value: &str) -> WithEnvVarSet {
+        let previous_env_value = env::var(key);
+        env::set_var(key, value);
+        WithEnvVarSet {
+            _previous_var_key: String::from(key),
+            _previous_var_value: previous_env_value,
+        }
+    }
+}
+#[cfg(unix)]
```
Comment (sylvestre) [inline src/uu/split/src/split.rs:362]: this is a lot of "cfg(unix)"
what about moving that into a specific platform file?
like
https://github.com/uutils/coreutils/tree/master/src/uu/tail/src/platform/

File: tests/by-util/test_split.rs (hunk 12 above)
Comment (sylvestre) [inline tests/by-util/test_split.rs:7]:
```suggestion
#[cfg(not(windows))]
use std::env;
```

File: src/uu/split/src/split.rs (hunk 17 above)
Comment (FelipeLema) [reply src/uu/split/src/split.rs:362]: sure, no problem

File: tests/by-util/test_split.rs [hunk 18]
```diff
@@ -156,3 +165,59 @@ fn test_split_additional_suffix() {
     assert_eq!(glob.count(), 2);
     assert_eq!(glob.collate(), at.read(name).into_bytes());
 }
+
+#[test]
+#[cfg(unix)]
```
Comment (sylvestre) [inline tests/by-util/test_split.rs:170]: One last request, could you please add a comment to explain why Windows doesn't work?

Comment (FelipeLema) [reply tests/by-util/test_split.rs:170]: sure... redundancy is always welcome

**Comment by sylvestre** [comment]
Well done :)

**Pull request merged**